enhancements/olm/platform-operators.md

```diff
@@ -0,0 +1,435 @@
+---
+title: platform-operators
+authors:
+  - "@timflannagan"
+reviewers:
+  - TBD
+approvers:
+  - "@joelanford"
+  - "@bparees"
+  - TBD
+api-approvers:
+  - TBD
+creation-date: 2022-06-30
+last-updated: 2022-07-06
+tracking-link:
+  - https://issues.redhat.com/browse/OLM-2232
+  - https://issues.redhat.com/browse/OLM-2513
+  - https://issues.redhat.com/browse/OLM-2605
+see-also:
+  - https://issues.redhat.com/browse/OLM-2447
+  - https://issues.redhat.com/browse/OLM-2570
+  - https://issues.redhat.com/browse/CORS-1873
+  - https://issues.redhat.com/browse/OCPPLAN-7589
+  - https://github.com/openshift/enhancements/blob/master/enhancements/update/cvo-techpreview-manifests.md
+replaces: []
+superseded-by: []
+---
+
+# Platform Operators
+
+## Summary
+
+The enhancement outlines a mechanism that enable support for the lifecycle of platform operators. A platform operator (PO) is an OLM-based cluster extension that can be installed during day 0 operations, and lifecycles with the cluster. This mechanism will ultimately be divided into phases, and this proposal covers the first phase is more detail.
+
+This first phase introduces a basic PO installation mechanism, and helps build the foundational layer for subsequent phases. Future phases may introduce a first-class PO management mechanism for the OLM component.
+
+## Motivation
+
+The motivation for platform operators is two-fold. The strategic direction for OpenShift is to make the core installation smaller by default and at the same time make it more flexible so that customers can tailor their installations to their use cases. This is what we’re calling “Composable OpenShift”.
+
+To make the core smaller, OpenShift is attempting to reduce the CVO-managed payload to the very minimum set of operators required to run the OpenShift control plane. From a control plane perspective, any non-essential operators can be installed after the cluster has been installed (or upgraded). However, from a customer standpoint, their business needs may dictate that these non-essential operators are still included in the lifecycle of cluster installations and upgrades. The Platform Operator (PO) concept will allow cluster administrators to choose which of these non-control-plane-essential operators to include in the cluster lifecycle.
+
+Secondly, some portion of the layered product operators in the OpenShift catalogs are conducive with inclusion in the cluster lifecycle (e.g. cert-manager, service binding operator, and the local storage operator). The Platform Operator concept will open the door for more of OLM’s optional operators to be included in (and impact) the cluster lifecycle. This will allow customers to further tailor their cluster installs and upgrades to their business needs.
+
+### User Stories
```

Comment:
i would love to see a user story about platform integrators who want to bundle their platform specific operators in a way that gives their users the best openshift performance out of the box on a new install.

Reply:
@elmiko I definitely think the new foundational components of OLM can help with that. The idea is that we'll eventually have:
1. a way to auto-detect more properties of the cluster (e.g. what cloud provider is it running on?) and make OLM aware of those things. 
2. a catalog with CCM operators that say "i am a CCM operator" and "i am only installable if cluster is in $cloud"
3. a cluster invariant constraint that requires "exactly one CCM operator must be installed" in the resolved set of operators

When OLM resolves all of those things together, it'll pick the CCM operator that is compatible with whatever cloud the cluster is on.

So perhaps the user story here is "As a platform integrator, I want Openshift to automatically install an appropriate CCM operator with PO semantics so that I can bundle platform specific operators in a way that gives my users the best integration and performance out of the box on a new install."

Reply:
i love to hear all of this, sounds exactly like what we need!

> So perhaps the user story here is "As a platform integrator, I want Openshift to automatically install an appropriate CCM operator with PO semantics so that I can bundle platform specific operators in a way that gives my users the best integration and performance out of the box on a new install."

while i do love the auto detection stuff, i think we need to always make sure that the users can control when, or if, the auto detection is enabled. i just want to make sure we always have the ability to explicitly state the operators in case the automatic detection gets unruly.

Reply:
It will absolutely be possible(eventually) to specify, in an install-config, "install these platform operators as part of the cluster install"

Until we have that behavior in the install-config/installer, it will also be possible to simply drop a "platformoperatorinstallrequest"(not the actual resource name) manifest into the generated manifests prior to starting the cluster creation (just like you can drop a subscription.yaml there today to install an olm operator during cluster bring up, it's just that if you do that, the installer won't wait for that operator to be ready)

Does that address your needs, @mhrivnak @dhellmann @elmiko ?

Reply:
> Does that address your needs, @mhrivnak @dhellmann @elmiko ?

absolutely.

i just want to make sure we are covering it, thanks =)

Reply:
My main concern wasn't necessarily that we have a way to install as part of day 0, since I thought that was always part of the plan. Something in the doc or thread above made me worried that the list of those platform operators would only be coming from a single catalog source that we manage, and if we assume that long-term that will be a problem. It's fine if the answer for installing POs from outside of our catalog involves a more complicated manifest workflow than just adding a name to the install-config, as long as there's some way to do it.

Reply:
> Something in the doc or thread above made me worried that the list of those platform operators would only be coming from a single catalog source that we manage

yes i think this EP implied/stated that at some point. I agree that's an unacceptable limitation for what we ultimately deliver here (whether it's an acceptable limitation in early phases or not)

Reply:
It's fine with me if it's a limitation of an early phase as long as it's not a limitation by the time we GA.

Reply:
Yeah, restricting to a single, known CatalogSource that already contains the initial set of operators we care about (e.g. cert-manager, service-binding-operators, etc.) is just to make it easier to onboard this new functionality in the short term, while we iron out the phase 1 design in the background. I would doubt we see this same restriction pop up during phase 1.

Reply:
> these user stories are missing the why/goal:
> 
> As a `$role` I want to do `$thing` so i can accomplish `$goal`

This is still true, can we make sure to update the user stories to include the motivation as to why a user wants to perform each action

Reply:
Updated the user stories in https://github.com/openshift/enhancements/pull/1170/commits/685968c97f2048b2f3eaee3ee8b84553210d60c2. There might be some personas missing there, but those are all scoped to phase 0.